core/src/main/scala/kafka/server/metadata/QuotaMetadataProcessor.scala

```diff
@@ -0,0 +1,190 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.server.metadata
+
+import kafka.network.ConnectionQuotas
+import kafka.server.ConfigEntityName
+import kafka.server.QuotaFactory.QuotaManagers
+import kafka.utils.Logging
+import org.apache.kafka.common.config.internals.QuotaConfigs
+import org.apache.kafka.common.metadata.QuotaRecord
+import org.apache.kafka.common.metrics.Quota
+import org.apache.kafka.common.quota.ClientQuotaEntity
+import org.apache.kafka.common.utils.Sanitizer
+
+import java.net.{InetAddress, UnknownHostException}
+import scala.collection.mutable
+
+
+// A strict hierarchy of entities that we support
+sealed trait QuotaEntity
+case class IpEntity(ip: String) extends QuotaEntity
+case object DefaultIpEntity extends QuotaEntity
+case class UserEntity(user: String) extends QuotaEntity
+case object DefaultUserEntity extends QuotaEntity
+case class ClientIdEntity(clientId: String) extends QuotaEntity
+case object DefaultClientIdEntity extends QuotaEntity
+case class ExplicitUserExplicitClientIdEntity(user: String, clientId: String) extends QuotaEntity
+case class ExplicitUserDefaultClientIdEntity(user: String) extends QuotaEntity
+case class DefaultUserExplicitClientIdEntity(clientId: String) extends QuotaEntity
+case object DefaultUserDefaultClientIdEntity extends QuotaEntity
+
+/**
+ * Watch for changes to quotas in the metadata log and update quota managers and cache as necessary
+ */
+class QuotaMetadataProcessor(private[metadata] val quotaManagers: QuotaManagers,
+                             private[metadata] val connectionQuotas: ConnectionQuotas,
+                             private[metadata] val quotaCache: QuotaCache) extends BrokerMetadataProcessor with Logging {
+
+  override def process(event: BrokerMetadataEvent): Unit = {
+    event match {
+      case MetadataLogEvent(apiMessages, _) =>
+        apiMessages.forEach {
+          case record: QuotaRecord => handleQuotaRecord(record)
+          case _ => // We only care about quota records
+        }
+      case _ => // We only care about metadata events
+    }
+  }
+
+  private[server] def handleQuotaRecord(quotaRecord: QuotaRecord): Unit = {
```

Comment:
What is your thinking behind all of this fine-grain visibility control? Who not make them all `public` the default in Scala?

Reply:
I guess I tend to make things private or package-private out of habit. Since this class is a processor/callback type of thing, I suppose visibility doesn't matter all that much.